File: week4&5/Ariling/prgs-파괴되지 않은 건물/main.js
function solution(board, skill) {
  var answer = 0;
  // 각 칸의 값들을 한꺼번에 저장해두기
  const arr = Array(board.length + 1)
    .fill()
    .map(() => Array(board[0].length + 1).fill(0));
  for (const [type, r1, c1, r2, c2, degree] of skill) {
    // 내구도 낮추냐 높이냐를 type으로 결정한다.
    const degrees = type === 1 ? -degree : degree;
    // 누적 범위를 구할 때 다음과 같이 한다
    arr[r1][c1] += degrees;
    arr[r1][c2 + 1] -= degrees;
    arr[r2 + 1][c1] -= degrees;
    arr[r2 + 1][c2 + 1] += degrees;
  }
  // 2차원 배열은 이런식으로 누적을 해줘야 한다고 한다...
  for (let i = 0; i < board.length; i++) {
    for (let j = 1; j < board[0].length; j++) {
      arr[i][j] += arr[i][j - 1];
    }
  }
  for (let j = 0; j < board.length; j++) {
    for (let i = 1; i < board[0].length; i++) {
      arr[i][j] += arr[i - 1][j];
    }
  }
  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[0].length; j++) {
      board[i][j] += arr[i][j];
      if (board[i][j] > 0) answer++;
    }
  }
  return answer;
}

File: week4&5/Ariling/prgs-파괴되지 않은 건물/main.py
def solution(board, skill):
    # 한꺼번에 계산해야 한다. 이거 for문하면 망한다를 직감적으로 느낌
    answer = 0
    arr = [[0] * (len(board[0])+1) for _ in range(len(board)+1)]
    for type, r1, c1, r2, c2, degree in skill:
        # 2차원 누적합은 이렇게 하는 것! 그리고 삼항연산자도 파이썬은 이렇게 한다.
        damage = -degree if type == 1 else degree
        arr[r1][c1] += damage
        arr[r1][c2+1] -= damage
        arr[r2+1][c1] -= damage
        arr[r2+1][c2+1] += damage
    for i in range(len(arr)):
        for j in range(1, len(arr[0])):
            arr[i][j] += arr[i][j-1]
    for i in range(len(arr[0])):
        for j in range(1, len(arr)):
            arr[j][i] += arr[j-1][i]
    for i in range(len(board)):
        for j in range(len(board[0])):
            board[i][j] += arr[i][j]
            answer = answer+1 if board[i][j] > 0 else answer
    return answer